src/server/pps/server/api_server.go

```diff
@@ -3624,3 +3662,111 @@ func newMessageFilterFunc(jqFilter string, projects []*pfs.Project) (func(contex
+	}
+	return &pps.GetClusterDefaultsResponse{ClusterDefaults: &clusterDefaults}, nil
+}
+
+// jsonMergePatch merges a JSON patch in string form with a JSON target, also in
+// string form.
+func jsonMergePatch(target, patch string) (string, error) {
+	var targetObject, patchObject any
+	if err := json.Unmarshal([]byte(target), &targetObject); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal target JSON")
+	}
+	if err := json.Unmarshal([]byte(patch), &patchObject); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal patch JSON")
+	}
+	result, err := json.Marshal(mergePatch(targetObject, patchObject))
+	if err != nil {
+		return "", errors.Wrap(err, "could not marshal merge patch result")
+	}
+	return string(result), nil
+}
+
+// mergePatch implements the RFC 7396 algorithm.  To quote the RFC “If the patch
+// is anything other than an object, the result will always be to replace the
+// entire target with the entire patch.  Also, it is not possible to patch part
+// of a target that is not an object, such as to replace just some of the values
+// in an array.”  If the patch _is_ an object, then non-null values replace
+// target values, and null values delete target values.
+func mergePatch(target, patch any) any {
+	switch patch := patch.(type) {
+	case map[string]any:
+		var targetMap map[string]any
+		switch t := target.(type) {
+		case map[string]any:
+			targetMap = t
+		default:
+			targetMap = make(map[string]any)
+		}
+		for name, value := range patch {
+			if value == nil {
+				delete(targetMap, name)
+			} else {
+				targetMap[name] = mergePatch(targetMap[name], value)
+			}
+		}
+		return targetMap
+	default:
+		return patch
+	}
+}
+
+func badRequest(ctx context.Context, msg string, violations []*errdetails.BadRequest_FieldViolation) error {
+	s, err := status.New(codes.InvalidArgument, msg).WithDetails(&errdetails.BadRequest{
+		FieldViolations: violations,
+	})
+	if err != nil {
+		log.Error(ctx, "could not add bad-request details", zap.Error(err))
+		s = status.New(codes.Internal, "could not add bad-request details")
+	}
+	return s.Err()
+}
+
+func unknownError(ctx context.Context, msg string, err error) error {
+	var stack []string
+	errors.ForEachStackFrame(err, func(f errors.Frame) {
+		stack = append(stack, fmt.Sprintf("%s:%d (%n)", f, f, f))
+	})
+	s, err := status.Newf(codes.Unknown, "unknown error: %s", msg).WithDetails(&errdetails.DebugInfo{
+		StackEntries: stack,
+		Detail:       err.Error(),
+	})
+	if err != nil {
+		log.Error(ctx, "could not add debug info details", zap.Error(err))
+		s = status.New(codes.Internal, "could not add unknown-error details")
+	}
+	return s.Err()
+}
+
+func (a *apiServer) SetClusterDefaults(ctx context.Context, req *pps.SetClusterDefaultsRequest) (*pps.SetClusterDefaultsResponse, error) {
+	var (
+		cd  = req.GetClusterDefaults()
+		crp pps.CreatePipelineRequest
+	)
+	if err := protojson.Unmarshal([]byte(cd.GetCreatePipelineRequestJson()), &crp); err != nil {
```

Comment:
I wonder if CPRJ wants to be `bytes` and not `string`?  I am not sure which is better.  On the one hand, it's opaque data.  On the other hand, JSON can only be utf-8, utf-16, or utf-32, and string is always `utf-8`, so I suppose that matches up well.  (What will languages with utf-16 strings do here?  I guess converting to utf-8 or base64-encoded utf-16 is about equally complicated.)

Reply:
Hmmm, you raise a good point. Languages with UTF-16 strings can always encode in UTF-8, of course. OTOH, we are storing the JSON the user submitted — does re-encoding still count as ‘the JSON the user submitted’?